src/coredump/coredump.c

```diff
@@ -375,39 +376,12 @@ static int save_external_coredump(
 
         (void) mkdir_p_label("/var/lib/systemd/coredump", 0755);
 
```

Comment:
hmm, so as I understand you now compress first, and then uncompress again, i.e. you trade CPU usage against disk space use.

This sounds surprising to me, I'd assume CPU is generally more scarce than disk space.

I would have always written things to disk first uncompressed as before, and if we see the size limit is hit we abort the attempt, compress what we have and write that out and don't do backtrace extraczion.

Reply:
There's two issues with that:

First, the storage location might be on a tmpfs (it is, in my case), so the size limit is not the one from the volume, but the amount of memory the kernel charges coredump's cgroup - and then you get zapped if you go over the limit, there's not really a way to fall back. Not even MemoryHigh saves you here - it will happily go over it and keep writing to the tmpfs, until MemoryMax is hit.

Then, even if there was a way around this, the fd we are reading is stdin so we can't unwind and start over to compress from the start.

With this change, if you don't want to pay the CPU cost you can just set `ProcessSizeMax=0`, and then you don't get a backtrace and you don't pay the decompression cost.

Reply:
> This sounds surprising to me, I'd assume CPU is generally more scarce than disk space.

Key thing is memory is scarcest - I can trade CPU with time, but I cannot conjure memory that doesn't exist :-/ Yes using memory-backed storage is a choice, but it is not that uncommon: think ephemeral systems and so on.

Reply:
But that sounds awfully specific in the usecase no? i.e. a memory constrained system with plenty CPU. I'd still believe that the opposite situation, where CPU is constrained but disk space is plenty is more common.

maybe try to be smart here? i.e. check if the storage we pick has plenty free space in which case we write out uncompressed first as beofre, and compress later, or if there' little space we compress first and uncompress again later?

(And maybe even if we detect that if only very very little space is free we don't write out at all, but just say "Sorry, no disk space"?)

Reply:
The problem is that the remaining space in the tmpfs on /var doesn't tell me much - it's the cgroup limit that is the problem. The tmpfs is shared, but in terms of cgroup, the kernel assigns the file size to the memory accounting of the process writing to it - which I guess makes sense.

I can't even know how big the uncompressed core is going to be, since it's from STDIN - I can only find out after either compressing it, or after writing it uncompressed. So what would I check against?

Reply:
i didn't know tmpfs was charged against memory cgroup. i knew that memfd was charged against it, but tmpfs I wasn't aware. Are you sure?

can't we determine the memory limit too? i mean, the cgroup attributes are easily accessible...

Isn't it kind of a broken setup to set up tmpfs with a huge size limit but not actually allowing people to write to it that much? maybe lower the size limit then? mount a separate tmpfs on /var/lib/coredump with a smaller size?

Reply:
but you just wrote it to disk, so you could read it from there?

Reply:
But it's truncated, no? So I can only compress that part, losing the rest

Reply:
hmm, you still have the pipe from the kernel open, where you can read the rest. i.e. you start reading the data from the pipe, and write it uncompressed to disk. Eventually you notice that this fills up too much, so you take the data you already wrote, and rewrite it – this time compressed. Then you delete the uncompressed version, and then you continue reading the data from the pipe and compress that part on-the-fly.

Reply:
The problem is that way a large amount of memory is still used, and the point where you notice it's "too much" might be too late and not have usable space anymore, as you cannot know in advance how well a core compresses. So the risk of losing cores is still high, which I need to minimize as it's really the worst case scenario.

What I am really looking for is a way to trade CPU usage for memory - but I understand this might not be desirable in all cases. So, how about the following implementation that I just pushed: the `Compress` option is now tristate. Default behavior is unchanged. If instead of `yes` it's set to `on-the-fly`, the new behavior I am after is used instead.

Is this acceptable as a solution?

Reply:
> The problem is that way a large amount of memory is still used, and the point where you notice it's "too much" might be too late and not have usable space anymore, as you cannot know in advance how well a core compresses. So the risk of losing cores is still high, which I need to minimize as it's really the worst case scenario.

Is this really an issue? I mean, we can be pretty sure that today's compression algorithms when they detect they don't compress data they write the data as-is, i.e. the ultimately written data will be a bunch of bytes for the header, and otherwise the exact uncompressed data size. The header is going to be less than 1K I figure, even for the fanciest of algorithms (simply because compression algorithms tent to be careful about not wasting to much). 

This if you'd normally put your coredump size limit for the coredumps to retain to value "x" then you could start writing out the coredump uncompressed until you reach limit y=(x-1024)/2. If you hit that limit you immediately stop. Instead you now compress your y bytes you already wrote to disk. The size of this will be y+1024 at most. Thus what you take up on disk will be y+y+1024 (i.e. one uncompressed version, one compressed version plus overhead for the header), and that's == x, hence you still didn't hit the overall maximum limit. Now you remove the uncompressed version, so you suddenly have at least y bytes free again (and most likely a lot more, because compression tends to work).

That should be simple enough and "just work" for everyone, no?

I am a bit conservative with the "on-the-fly" thing, because to me that implements an implementation detail that better shouldn't be exposed in the configuration, because people shouldn't pick algorithms, but metrics and preference we then use to pick and tune an algorithm.

In that sense it might be OK to expose this as a new option Priorize=process|storage or something like that. i.e. that declares what should be more important: that the stuff is analyzed the most efficient way or that it hits the disk in compressed form. i.e. Priority=process would mean: priorize backtrace generation, i.e. make things quick at the chance that later on disk space might not suffice (or the system failed) to write the image. Priority=storage would mean: priorize storage (and this means compression), even if that makes things slower, since we need to decompress things again.

But I still would prefer if we'd not make this configurable and instead try to make things just work as describe above, by picking a safer, lower threshold when temporarily storing uncompressed, and the configured threshold only when finally storing the compressed coredump.

Reply:
I agree with @poettering here: making the code smart and simply honouring ProcessSizeMax= as far as possible without explicit configuration seems better. (I unset `needs-review`, since the commits are older than the comments here.)

Reply:
Working on this - the tricky part is figuring out the exact limit, because in my case I don't simply get an out of space error, but oom-killer unceremoniously blasts me out of the sky. I'm checking whether getting MemoryMax out of the unit and its slice will work.

Reply:
Implemented as suggested - need to figure out the best way to get MemoryMax for the unit and slice it belongs to, for now I'm querying systemd via dbus.
Also I've not tested this in production, only locally - if it turns out it's not as reliable, I might come back with an additional explicit option in the future - hopefully it won't be necessary. PTAL!